src/android/java/io/jxcore/node/BtConnectorHelper.java

```diff
@@ -68,119 +74,77 @@ public boolean isRunning(){
         }
     }
     public void Stop(){
-        if(mBTConnector != null){
-            mBTConnector.Stop();
-            mBTConnector = null;
+
+        BTConnector tmpCon = mBTConnector;
+        mBTConnector = null;
+        if(tmpCon != null){
+            tmpCon.Stop();
         }
 
-        if (mServerSocketList != null) {
-            for(int i = 0; i < mServerSocketList.size(); i++) {
-                BtToServerSocket tmp = mServerSocketList.get(i);
-                if (tmp != null) {
-                    print_debug("Disconnect:::Stop : mBtToServerSocket :" + tmp.getName());
-                    tmp.Stop();
-                    tmp = null;
-                    mServerSocketList.set(i, tmp);
-                }
-            }
+        //disconnect all incoming connections
+        DisconnectIncomingConnections();
+
+        // disconnect outgoing connection
+        Disconnect("");
+    }
 
-            mServerSocketList.clear();
+    // we only cut off our outgoing connections, incoming ones are cut off from the other end.
+    // if we want to cut off whole communications, we'll do Stop
+    public boolean Disconnect(String peerId){
+        BtToRequestSocket tmpSoc = mBtToRequestSocket;
+        if(tmpSoc == null) {
+            return false;
         }
 
-        if(mBtToRequestSocket != null) {
-            print_debug("Disconnect:::Stop : mBtToRequestSocket");
-            mBtToRequestSocket.Stop();
+        String currentpeerId = tmpSoc.GetPeerId();
+        print_debug("Disconnect : " + peerId + ", current request : " + currentpeerId);
+        if (peerId.length() == 0 || peerId.equalsIgnoreCase(currentpeerId)) {
+            print_debug("Disconnect:::Stop :" + currentpeerId);
+
+            tmpSoc.Stop();
             mBtToRequestSocket = null;
+
+            return true;
+
         }
+        return false;
     }
 
-    public boolean Disconnect(String peerId){
+    //test time implementation to simulate 'peer disappearing'
+    public boolean  DisconnectIncomingConnections() {
 
-        boolean ret = false;
+        if (mServerSocketList == null) {
+            return false;
+        }
 
-// we only cut off our outgoing connections, incoming ones are cut off from the other end.
-// if we want to cut off whole communications, we'll do Stop
-        if(mBtToRequestSocket != null) {
-            String currentpeerId = mBtToRequestSocket.GetPeerId();
-            print_debug("Disconnect : " + peerId + ", current request : " + currentpeerId);
-            if(peerId.length() == 0 || peerId.equalsIgnoreCase(currentpeerId)) {
-                print_debug("Disconnect:::Stop :" + currentpeerId);
-                mBtToRequestSocket.Stop();
-                mBtToRequestSocket = null;
-                ret = true;
+        for (int i = 0; i < mServerSocketList.size(); i++) {
```

Comment:
It is super unsafe to iterate on a thread safe object using a counter since the membership can change during the iteration. You have to model this stuff in your head. On each method call you have to ask yourself "what happens if we have multiple simultaneous calls?" If multiple people can work on mServerSocketList at the same time then certainly while we are iterating on its membership that membership can change! Also whenever using a new class you should always read its docs. The docs on CopyOnWriteArrayList (see [here](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html)) explicitly calls out the fact that it uses a snapshot style iterator to handle exactly this problem. So what you need to do is to call either iterator() or listIterator() in order to get a thread safe snap shot of the list's contents and iterate on that.

Also using that iterator should simplify the code below since, for example, you don't need to check for null. You should never be putting nulls into the list. Instead when you want to remove a member you should call remove. Keep in mind, btw, that you should NEVER remove by index. It's never safe. I'm shocked they even offer it as an option.

In fact, if I were you, since order doesn't matter to either lastAvailableList or mServerSocketList I would urge you to consider changing both variables to be of type CopyOnWriteArraySet. This will remove most of the really unsafe methods and make it easier to write correct code.